### ISSUE TITLE
Support for composite types?

### ISSUE THREAD

**Issue by bitemyapp** [yesodweb/persistent#787]
This might be `persistent-postgresql` rather than `persistent` but it seems like composite types aren't supported at the moment. This thought was prompted by: https://joncfoo.gitlab.io/posts/2016-06-17-composite-types-with-postgresql.html

I think there's a way to do this by making your own and instances so this is mostly me just musing about formal support.

**Comment by MaxGabriel** [comment]
I'm actually not sure you could make your own instances. In the implementation of `PersistField`, you'd presumably use `PersistDbSpecific bs` to store something like `"row(null,null,'city','state','zip')"`. But then in `persistent-postgres`, `PersistDBSpecific` is made into a field using `Unknown`:

```
    toField (P (PersistDbSpecific s))  = PGTF.toField (Unknown s)
```

whose `ToField` instance is:

```
instance PGTF.ToField Unknown where
    toField (Unknown a) = PGTF.Escape a
```

So I think that would happen is that all `"row(null,null,'city','state','zip')"` would be escaped incorrectly and couldn't be inserted into the database.

**Comment by bitemyapp** [comment]
@MaxGabriel the approach I was alluding to was nested products that would flatten the representation and repeat the columns across the tables that had them in common. Is there a reason doing that with a nested product and custom instances wouldn't work? I'll probably sketch it out into something unambiguous regardless but I'd like to know if I'm missing something here as I am pretty sure I have done more complicated things than that w/ custom Persistent record types.

**Comment by MaxGabriel** [comment]
You mean like the Haskell type would be `Address` and that would map to multiple columns in the database?

**Comment by bitemyapp** [comment]
@MaxGabriel yeah I'm imagining would get manually lifted up by the enclosing type. Is that not possible in terms of the existing typeclasses? If so, that should be fixed first.

**Comment by MaxGabriel** [comment]
I didn't think that was possible but I could be wrong. That would be a super useful feature for me at work so I'd be very happy to be wrong about it

**Comment by bitemyapp** [comment]
@MaxGabriel okay, I may have missed something then, I'll look into it and report back. Thank you!
